Forms: Add file upload dropzone block

edanzer opened the pull request:
## Proposed changes:

- Refactors the file field block to contain label block + new dropzone block, built on top of global styles refactor:
    - https://github.com/Automattic/jetpack/pull/42890
- Based on earlier [PR 42980](https://github.com/Automattic/jetpack/pull/42980/commits/35d3db7a2f3968d58b18bbf6556e0668fc41f525)
  
###  **Before**, the block would use "group" block as its container

<img width="819" alt="Screenshot 2025-05-15 at 12 32 33" src="https://github.com/user-attachments/assets/e808370b-9d4f-4170-a9b2-9ab5de18d3dc" />
<img width="1113" alt="Screenshot 2025-05-15 at 12 32 57" src="https://github.com/user-attachments/assets/9c5290e0-0ee2-4eb1-9482-3844178a062c" />


###  **After**, the block uses separate "dropzone" block

<img width="819" alt="Screenshot 2025-05-15 at 12 24 09" src="https://github.com/user-attachments/assets/54b45299-ebb8-4a40-8f20-8a78a8f1261e" />
<img width="1117" alt="Screenshot 2025-05-15 at 12 23 20" src="https://github.com/user-attachments/assets/3f9f768c-89f8-46c2-89ab-4857c3476d37" />



### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None.

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Enable beta blocks by adding this to your environment:
    ```php
    add_filter( 'jetpack_blocks_variation', function () { return 'beta'; } );
    ```

* Add form with file upload field. 
* Confirm file upload and dropzone look right in editor. 
* Confirm file upload and dropzone look correct and work as expected on frontend. Add `jetpack-beta-blocks` sticker via Blog RC to allow file uploads to your site.
* Refresh the block editor page, and confirm the file upload and dropzone block still load as expected.

aaronrobertshaw committed:
Add base form field blocks

Introduces foundational form field blocks:
- jetpack/input
- jetpack/label
- jetpack/option
- jetpack/options

These blocks will be reused across all field types using inner blocks.

aaronrobertshaw committed:
changelog

aaronrobertshaw committed:
changelog

aaronrobertshaw committed:
Migrate form field blocks to use new base blocks

Updates legacy form field blocks to use shared inner blocks:
- jetpack/input
- jetpack/label
- jetpack/option
- jetpack/options

This commit also updates the registration of form blocks so the new
shared blocks are available in the editor.

aaronrobertshaw committed:
Fix CSS linting issues

aaronrobertshaw committed:
Fix issues with style syncing hook:

- Prevents adding bulk changes when syncing blocks to the undo stack
- Correctly determines the parent field block for option blocks to check
  if syncing is enabled
- Correctly find all option blocks regardless of whether they are a
  field's direct child, e.g. consent, or nested within an options block
  e.g. single/multiple choice fields.

aaronrobertshaw committed:
Make select dropdown field's edit component more robust

aaronrobertshaw committed:
Prevent style syncing when blocks are reused in previews

aaronrobertshaw committed:
Ensure label is editable after being emptied and placeholder shows

aaronrobertshaw committed:
First pass at updating form variations

aaronrobertshaw committed:
Quick clean up of form variations

aaronrobertshaw committed:
Pass required block toolbar control through to Option block for checkbox field

aaronrobertshaw committed:
Clean up form field and inner block attributes

aaronrobertshaw committed:
Remove lodash dependency from dropdown field edit component

aaronrobertshaw committed:
Remove remaining lodash uses in new or updated form blocks

aaronrobertshaw committed:
Remove TODO as Dropdown fields didn't use option styles

aaronrobertshaw committed:
Fix default styling for Consent field option label

aaronrobertshaw committed:
Clean up unrequired TODO item

aaronrobertshaw committed:
Clean some more TODOs

aaronrobertshaw committed:
Restore empty string defaults for input, option, and label blocks

aaronrobertshaw committed:
Revert use sync style hook to previous version retaining undo stack fix

aaronrobertshaw committed:
Fix typo

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

aaronrobertshaw committed:
Fix linting error in editor.scss

talldan committed:
Update form field style syncing (#43004)

* Refactor and update form field syncing.

- Use a parent react context provider as the source of truth instead of an
arbitrary sibling block.
- Perform a deep equality check before syncing to prevent against spurious
updates.

* When updating the block own attributes, mark the change as not persistent so that the user can undo

* Prime the previousOwnAttributes value so that we avoid updating syncedAttributes unnecessarily

* Update inline docs

* Prime the synced attributes when they are initially unset

* Fix syncing of styles across checkbox and consent options

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Add transforms for new inner blocks based fields

andrewserong committed:
Attempt to fix warnings about PhanUndeclaredTypeParameter for WP_Block

andrewserong committed:
Fix linting issue in class-contact-form.php

aaronrobertshaw committed:
Forms: Prevent errors when synced attribute context is not available (#43027)

* Forms: Fix error thrown in previews for synced form fields

* Clean up

* Rename the context type stuff

andrewserong committed:
For html_entity_decode, use ENT_COMPAT since we only care about double quotes for the purposes of JSON decoding

andrewserong committed:
Remove is_array and empty check as we now will always have at least one 'class' attribute on the extra_attrs array

aaronrobertshaw committed:
Clean up standalone checkbox inner blocks template

aaronrobertshaw committed:
Clean up dropdown field inner blocks template

aaronrobertshaw committed:
Add example definitions for field block types

aaronrobertshaw committed:
Allow default checked status for standalone option fields

aaronrobertshaw committed:
Fix choice field deprecation edge case

aaronrobertshaw committed:
Fix handling of required and width attributes for Textarea field.

aaronrobertshaw committed:
Fix label block's placeholder fallbacks

talldan committed:
Fix form undo redo issues (#43036)

* Fix undo stack issue with useFormWrapper

* Fix undo stack issue with text area

* Fix undo stack issue with duplicated useFormWrapper

* Fix undo stack issue with consent field block

* Fix undoing after selecting a block variation causing multiple undo steps

* Fix comment typos

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

* Remove `required` property from template

This is now provided to the child block through context

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

ramonjd committed:
Forms: Add PHP unit tests for Migrate form fields to inner blocks JETPACK-262  (#43002)

* inital commit - exploratory

* do_shortcode

* Add test_block_attributes_to_shortcode_attributes_with_styles

* add more simple tests

* remove unnecessary changes

* changelog

* LINTO!

* Adding WP namespaces

* test_gutenblock_render_field_checkbox

* Add test for trim

* Use optionsdata to presume that block_attributes_to_shortcode_attributes has taken place. Let's test that as a unit

* Adding classes and styles in attributes.

* block_attributes_to_shortcode_attributes tests

* text field which has input and label

* Phan complains about \DOMNode::getAttribute. Shut it up.

* Try to quiet Phan again by passing a WP_Block instance

* Attempt to fix a couple of the code coverage warnings

* Try to fix namespace again

* Try using the full namespace for the covers annotation instead of the relative name

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Forms: Add PHP unit tests for Migrate form fields to inner blocks JETPACK-262  (#43002)

* inital commit - exploratory

* do_shortcode

* Add test_block_attributes_to_shortcode_attributes_with_styles

* add more simple tests

* remove unnecessary changes

* changelog

* LINTO!

* Adding WP namespaces

* test_gutenblock_render_field_checkbox

* Add test for trim

* Use optionsdata to presume that block_attributes_to_shortcode_attributes has taken place. Let's test that as a unit

* Adding classes and styles in attributes.

* block_attributes_to_shortcode_attributes tests

* text field which has input and label

* Phan complains about \DOMNode::getAttribute. Shut it up.

* Try to quiet Phan again by passing a WP_Block instance

* Attempt to fix a couple of the code coverage warnings

* Try to fix namespace again

* Try using the full namespace for the covers annotation instead of the relative name

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

aaronrobertshaw committed:
Clean up TODO as the important style for Outlined form style can not be avoided

aaronrobertshaw committed:
Remove extraneous hook dependency

aaronrobertshaw committed:
Remove useFormWrapper audit TODO - all fields use it

aaronrobertshaw committed:
Remove TODOs already captured by issues

talldan committed:
Remove unused ids from consent, checkbox, single choice and multiple choice fields (#43055)

aaronrobertshaw committed:
Opt out of block gap spacing for Options block

This is an enhancement over trunk however the is some work required to
make this actually work, especially given the shortcode rendering on the
frontend. For the time being this support will explicitly be opted out
of. It can be revisited when the benefit justifies the additional
complexity and work.

aaronrobertshaw committed:
Fix saving and display of dropdown field placeholder

aaronrobertshaw committed:
Fix dropdown field placeholder default

This commit also updates the dropdown field input's styles to make it
easier to click when the placeholder has previously been cleared.

aaronrobertshaw committed:
Fix style linting error

ramonjd committed:
Form inner blocks: supplement existing PHP unit tests (#43076)

* Enhance unit tests for Contact_Form_Block by adding support checks

This commit updates the `test_register_child_blocks` method to include an optional parameter for expected block supports and introduces a new test method, `test_child_blocks_block_supports`, to verify the configuration of block supports for various Jetpack blocks. The data provider for the new test method is also added, ensuring comprehensive testing of block support attributes.

* Add unit test for gutenblock_render_field_radio method

This commit introduces a new test method, `test_gutenblock_gutenblock_render_field_radio`, to validate the rendering of the radio field block in the Contact Form Plugin. The test checks that the generated shortcode matches the expected output, ensuring proper functionality of the radio field block with various attributes and inner blocks.

* changelog

* remove duplicate tests. silly me!

* Apply suggestions from code review

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

* syncing server block supports with frontend

* try again

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Allow default checked status for standalone option fields (#43077)

talldan committed:
Fix date picker field incorrect classname (#43059)

* Ensure jetpack-field__input classname applies to date picker input instead of jetpack-field__date

* Remove explicit date type in appointment form variation

* Specify the field type for the date field in the block template

* Iron out inconsistency of date type for date picker inputs

* Tied up text fields array

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Revert "Remove deprecated covers annotations in line with #42902"

This reverts commit 38791a8459ac68bb654be2d724530e5473f218dd.

Restoring these in case they were in place for coverage requirements.

ramonjd committed:
Forms: move @covers to class level in PHP unit tests (#43081)

* Hoist @cover annotations to class level.

* changelog

talldan committed:
Fix field block examples (#43096)

* Fix field block examples being altered by useFormWrapper

useFormWrapper adds a form wrapper around any field that doesn't have one.

It wasn't preserving the field's own inner blocks, which became visible in the block
examples, where single or multiple choice fields have inner blocks but no wrapper.

This PR ensures the inner blocks are preserved by calling getBlocks to get the inner
blocks of the field and passing that as the inner blocks in the replaceBlocks call
within useFormWrapper.

* Small improvement to variable name

* Add missing useEffect deps

* Fix select/dropdown example

* Switch e2e test close welcome guide code to use preferences

edanzer committed:
Add dropzone block for file uploads

aaronrobertshaw committed:
Forms: Fix display of label and option fallback values in transform previews JETPACK-278 (#43100)

talldan committed:
Add fallback styles for dropdown block when css vars are not populated (#43101)

aaronrobertshaw committed:
Add missed cleaned up label fallback handling

aaronrobertshaw committed:
Fix rendering of legend as a label

- Ensure block class is added to legend element
- Add custom block classes and styles to legend
- Tweak default style specificity so global styles can apply

aaronrobertshaw committed:
Fix rendering of legend as a label

- Ensure block class is added to legend element
- Add custom block classes and styles to legend
- Tweak default style specificity so global styles can apply

aaronrobertshaw committed:
Remove redundant array check to make linter happy.

aaronrobertshaw committed:
Clean up dead code in shared JetpackFieldControls

andrewserong committed:
Try to fix metadata in doc comments warning

andrewserong committed:
Try updating PHP test attributes again, to hopefully please the linter

ramonjd committed:
Forms: ensure custom CSS in global styles applies correctly to checkbox and radio wrappers (#43184)

* Refactor class attribute handling in Contact_Form_Field

- Trimmed type variable to ensure consistent class generation.
- Added specific classes for 'checkbox-multiple' and 'radio' types.
- Included TODOs for potential improvements and testing.

* changelog

* Add block styles for checkbox-multiple and radio fields

- Registered new block styles for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio'.
- Updated class handling in Contact_Form_Field to ensure proper application of styles.
- Adjusted CSS selectors to reflect changes in class structure for button styles.

* Enhance class handling for checkbox-multiple and radio fields

- Updated Contact_Form_Field to apply block style classes to wrapper divs for checkbox-multiple and radio types.
- Introduced a new method to extract block style classes from the class attribute.
- Modified tests to validate the addition of block style classes in the rendering process.

* Refactor block style registration for checkbox-multiple and radio fields

- Consolidated block style registration for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio' into a single call.
- Updated comments to clarify the purpose of style registration and future improvements.
- Adjusted class handling in Contact_Form_Field to include new class naming conventions.

* Refactor block style application for checkbox-multiple and radio fields

- Streamlined the handling of block style classes for 'checkbox-multiple' and 'radio' types in Contact_Form_Field.
- Improved code readability by consolidating conditional checks for block style class assignment.
- Ensured consistent application of styles to wrapper divs for both field types.

* Enhance radio field rendering tests and block style application

- Added tests to validate rendering of radio fields with block style classes.
- Improved handling of block style classes for radio fields in Contact_Form_Field to ensure consistent application.
- Updated assertions in tests to reflect changes in class structure for radio fields.

* Remove outdated changelog entry for migrating form fields to inner blocks with multiple wrapper styles

* Enhance field wrapper class handling for contact form fields

- Introduced 'fieldwrapperclasses' attribute to allow custom wrapper classes for form fields.
- Updated rendering logic in Contact_Form_Field to utilize the new attribute for consistent styling.
- Refactored tests to validate the addition of 'fieldwrapperclasses' for various field types, ensuring proper application of block styles.

* Try fixing a linting issue with the DataProvider

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

andrewserong committed:
Add icons for input, label, option, and options blocks (#43270)

talldan committed:
Form Global Styles: Update CSS specificity to 0-1-0 for default editor styles (excluding outlined style) (#43125)

* Reduce specificity of text area styles to 0-1-0

* Update specificity of option block styles

* Fix option input not scaling with font size for single/multiple choice fields

* Fix font:inherit rule overriding global styles font family setting by reduce specificity

* Add dead code comments

ramonjd committed:
Forms: only apply `wp-` and `is-style-` classes to fields of new inner block types (#43268)

* Only apply the block style classes to the field wrapper if the field is one of the new inner block types.

* update tests

* Refactor get_block_style_classes to return field wrapper classes and simplify class handling

* Ensure there's no whitespace in unimgrated fields

talldan committed:
Fix missing checkmark for single choice options on frontend (#43301)

talldan committed:
Forms: Fix CSS specificity for compatibility with global styles on the frontend (#43299)

* Fix consent field label css specificity preventing global typography styles

* Fix checkbox/radio option font-weight css specificity preventing global styles

* Remove empty line and duplicate classname

* Fix mixed spaces/tabs indentation

* Fix checkbox/radio input not scaling with font size for single/multiple options fields

* Fix indentation

* Fix button style multiple choice checkbox not receiving global styles font color due to css specificity

* Fix single choice option not receiving text color due to styles not applying to correct element

* Reinstate missing color style

* Move option block style classnames to parent wrapper on frontend

* Make option label font weight inherited from parent block wrapper, add some back compat for default css styles

* Fix tests

* Move option styles to parent block level element

simison commented:
Rebased against branch [`update/migrate-form-fields-to-inner-blocks`](https://github.com/Automattic/jetpack/tree/update/migrate-form-fields-to-inner-blocks) (which is worked on PR https://github.com/Automattic/jetpack/pull/43114)

ramonjd committed:
Forms: remove color gradients block supports (#43427)

* Forms: Update gradient property naming in contact form block and tests to ensure consistency

* Forms: Update gradient property naming to 'gradients' for consistency in input and option blocks

* Fix label typo

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
[WIP] Forms: migrate Outlined form style for inner block based fields (#43327)

* first commit

* backend mania

* Removing dividing two top of label. Can reinstate

* fix number conversion

* Match legacy form behaviour when deriving px values for CSS vars

* color, typography

* Adjust clip path to cater for 1px borders

* Implement outline style support for form fields by integrating useFormStyleOutlineClassesAndStyles hook across various blocks. Update CSS variables and styles for outlined and animated form styles in the contact form backend and frontend components.

* Cannot fix overlap right now

* Trying to fix select, options, etc

* desperately trying to assign the right values to css vars

* get background working

* path to syncing options

* use lodash for merging

* global styles specificity

* Remove isSynced from hook and related code

* Try only using hook on label

* Inject CSS vars for input as well as label so that it can calculate correct padding

* Call it a jetpack contact form notch width

* Fix overflowing background when a narrow border-width is set

* Fix input block placeholders not selectable/visible when outlined style is active

* Fix animated style label and input offset

* Animated style: fix label and input offset when using distinct border left style in editor

* Fix trailing notch background color

* Inputs should take border radius styles so they match the notched border. Fix global style derivation for borders

* Fix incorrect block names

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

* Improve support for split border radii in outlined form style

* Remove min width for notch

* This commit syncs jetpack/options with jetpack/input

* abstract get inner block hook

* esc notched attrs

* Check $legacy values before assignment

* move css vars to inner elems

* Apply the background to the notch elements again instead of the input to ensure border aligns with background. Remove the z-index rule so that the input/placeholder is visible and can be edited

* Adding padding where there's a border

Enhance contact form styles and structure by adding support for animated styles, adjusting border colors, and refining fieldset rendering logic.

* Add filler notch element on frontend

* Fix flex shrink/grow of notch elements with new filler notch element

* Limit padding of inputs to support a wider range of border radii units

This helps prevent situations where the calculated padding can end up
being a huge number when units like % are used. 100% might make sense for
a border radius, but that value used for a padding is huge. Cap the padding
at 100px to match trunk's max border radius value.

* Ensure dropdowns have same styles for input as other fields, DRY up

* Updated rendering logic in the contact form fields to differentiate between outlined and animated styles. Introduced a new helper function to manage outlined style attributes and adjusted CSS to ensure proper styling for outlined fields. Cleaned up related code for clarity and maintainability.

* Ensure default padding for outlined styles in inputs

* Refactor contact form styles to ensure global styles can override default border and padding settings. Updated class logic for option groups to conditionally apply border styles based on attributes, enhancing flexibility and maintainability.

* Account for split border radius when calculating css var on the frontend

* Set a max width for any elements using --jetpack--contact-form--notch-width of 100px

* Update tests with new shortcode attributes and styles for outline and options

* Switch inputs and labels for the outlined and animated style so that label is above input

* Remove background on inputs for outlined style, background is being applied to labels instead

* Add filler notch element to match editor styles

* Add max-width for trailing notch element

* Fix border radius css var not picking up single values with a unit

* Remove border radius and clip path on inputs for outlined style (border radius is applied on the label, clip path does nothing)

* Update tests with new attributes and outline HTML

* Default white background in the editor

* Ensure the Button variation on the Multiple Choice (Checkbox) block inherits the current color from the option block.

* Add background color support for outlined style in contact form inputs and update related tests

* Fix outlined border radius for checkbox multiple

* Fix css border-radius var calculation

* Fix checkbox multiple radius for outlined style

* Small tidy up

* Enhance outline styles handling in contact form fields by adding form style parameter and refining border radius calculations. Update CSS to include z-index for animated labels.

* Apply outlined border radius style to radio fields to match other fields

* Remove spaces in inline styles

* Select arrow position, default padding for has-background

* Padding magic on the options in outlined

* Add font size adjustment for legend in outlined contact form style

* Fix position of legend and right padding for options

* Remove export of constant

* Combine is_outlined `if` statements for checkbox multiple

* Combine is_outlined `if` statements for radio

* Rename useFormStyleOutlineClassesAndStyles to useVariationStyleProperties

* Only memoize based on changes to block type when merging global styles

* Add documentation to useVariationStyleProperties

* Small tidy up, rename a function and add docs

* Remove commented out css declarations

* Remove dead code. Context is unused. Anywhere. Go look if you don't believe me.

---------

Co-authored-by: Daniel Richards <daniel.richards@automattic.com>

andrewserong committed:
Add esc_attr calls to render_radio_field (#43488)

ramonjd committed:
Reinstate default padding where there's a border. Add option chaining to innerBlock look up as trunk forms won't have it.

talldan committed:
Use correct block name in get_ouline_styles (#43489)

ramonjd committed:
Forms: standardize animated variation style paddings (#43469)

* Update padding for animated style

* Fix updated function

* Rename get_outline_styles

* check borderradius

* Calculate border widths with all sides via CSS vars

* Fix text area top position

* Reinstate default (trunk) padding for animated labels etc

* Default value of null on border sizes

* Remove dupe CSS

* JS Doc :success:

* Add legacy CSS var as default

* Use the `borderwidth` field attribute as the legacy legacy border width value

* Explain use of `borderwidth` value

* Update projects/packages/forms/src/contact-form/class-contact-form-field.php

Use var `$border_width_attribute`

* Add bottom width CSS var to backend, and use border sides in the notched CSS

* Remove styles that we're overriding

* Switch from !empty to isset for legacy borderwidth attribute check

* Do empty/zero check on legacy border widths

---------

Co-authored-by: Daniel Richards <daniel.richards@automattic.com>

enejb committed:
unset the flex property so that blocks works more as expected.

talldan committed:
Forms: Fix block instance border color not overriding global styles border color with outlined style active (on the frontend) (#43532)

* Allow block instance border color to override global styles border color

* Remove extra space in class name

andrewserong commented:
Just checking — this PR doesn't include any transforms / deprecations to migrate ad hoc styles to use the block supports, as other blocks have done in the inner blocks refactor.

Will the non-inner-blocks version of File Upload ever be used by real users in production? If so, then we might need to add some handling here. If not, then I imagine we might not need any deprecations to migrate the styles, but just thought I'd double check just in case.

ramonjd committed:
Revert input + option syncing. Bad!

simison commented:
>  Will the non-inner-blocks version of File Upload ever be used by real users in production? If so, then we might need to add some handling here. If not, then I imagine we might not need any deprecations to migrate the styles, but just thought I'd double check just in case.

Thanks for checking! We indeed won't want to release the non-inner-blocks version to customers and rather wait for any time needed to release the refactor and the new feature together.

talldan committed:
Forms: Fix typography input styles applying to labels when outline style is active (on frontend) (#43534)

* Fix rogue typography attributes from input applying to label

* Fix bad rebase

* Update tests

ramonjd committed:
For classic themes, do not destructure theme config objects that don't exist.

andrewserong committed:
Forms: Fix custom classnames support in inner blocks migration (#43564)

enejb committed:
Make the form more accessibe.

enejb committed:
Add is-layout-constrained class so that things center more as expected.

ramonjd committed:
Forms: fix custom class handling in contact form plugin (#43587)

* Forms: Enhance class handling in contact form plugin and tests

- Updated class handling in `class-contact-form-plugin.php` to filter empty block style classes.
- Modified test cases in `Contact_Form_Plugin_Test.php` and `Contact_Form_Test.php` to include additional attributes and ensure proper class wrapping for various field types.
- Improved assertions to validate class attributes in rendered output.

* Make innerblocks optional. Just for linting

talldan committed:
Forms: Fix outline style label font size (#43590)

* Fix text input-style field label font size when outlined style is active.

* Remove scaling using variable

* Fix scaling of multiple option fields when using outlined style

* Tidy up styles

* Wrap labels

* Fix outlined label font size and animation

* Fix animation by ensuring top is a valid value

* Remove will-change: transform on frontend too. Transform usually does not change.

talldan committed:
Forms: Fix outline style label font size (#43590)

* Fix text input-style field label font size when outlined style is active.

* Remove scaling using variable

* Fix scaling of multiple option fields when using outlined style

* Tidy up styles

* Wrap labels

* Fix outlined label font size and animation

* Fix animation by ensuring top is a valid value

* Remove will-change: transform on frontend too. Transform usually does not change.

andrewserong commented on projects/packages/forms/src/blocks/dropzone/index.js:
```diff
@@ -0,0 +1,118 @@
+			default: {
+				type: 'constrained',
+			},
+		},
```
Does this block need all the different configurations of layout, or is the most common case that folks will center whatever content is within the dropzone? I.e. just some ideas:

* If we always expect `constrained`, then we could have that as the default, but switch off `allowSwitching`?
* Or would it work with `flex` layout instead of `constrained`, so that the container would work a bit more like the Stack variation of the Group block, or does it need all the different kinds of configuration? (Also, I wasn't sure if `constrained` is used so that the wrapper defaults to full width 🤔)
* Often `flex` can be useful as we get things like orientation for free, without having to expose the layout switcher, so we could still support horizontal and vertical configurations.
* So, it could be possible to set this with a default value that works for the majority of cases. Then, we could always follow up and add more options if folks find they need it?
* Also, worth testing with Classic themes that don't use all the layout options, to make sure everything's still working correctly — i.e. Classic themes that don't opt into appearance controls may not have any styles associated with `is-layout-constrained`.

For context, @annezazu flagged that exposing the layout switcher can be confusing for folks (p1748006710383329-slack-C02A76B714Z) and this comment represents my quick search to see where we might be exposing that layout switcher 🙂

andrewserong commented on projects/packages/forms/src/blocks/dropzone/index.js:
```diff
@@ -0,0 +1,118 @@
+		// Mimic the layout settings of the core Group block.
+		layout: {
+			type: 'constrained',
+			allowSwitching: true,
```
Why is layout switching enabled? I don't think any core blocks use it, as the layout switching component never really got to a good point of usability.

enejb replied:
Good idea. I am still trying to understand what is the best "layout" but so far in my testing the "constrained" one seems to work the best. 

I don't really see the need to be able to switch bettween the different onces.

andrewserong replied:
Mostly curious, but how come constrained is working the best, and what are the trade-offs between constrained, flow/default and flex so far in your testing?

With constrained layout, child blocks can typically set wide and full width alignments, and at first glance, that doesn't seem to be the use case here.

The choice of layout type typically comes down to the kind of content that folks are managing. I.e. for Buttons or Social Icons, flex is a natural choice, but I'm curious about the kinds of blocks or layouts we're expecting/hoping folks will lay out within the dropzone. Maybe some ideas or examples of patterns of content within the dropzone might help us figure out which layout might work best here 🤔

simison replied:
Just basically paragraphs, buttons, dividers, and images. Most likely centered and as a column, and might get left/right aligned.

We're now allowing group/columns and some other layout blocks too but I recon we'll remove access to those for now to simplify based on feedback we got at the demo, and add back if customers ask for them.

andrewserong commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -1115,6 +1115,16 @@ private function render_file_field( $id, $label, $class, $required, $required_fi
+			$processor = new \WP_HTML_Processor( html_entity_decode( $this->content, ENT_COMPAT, 'UTF-8' ) );
+			while ( $processor->next_tag() ) {
+				if ( $processor->has_class( 'wp-block-jetpack-dropzone' ) ) {
+					$processor->add_class( 'is-layout-constrained' );
```
I've already left a couple of other comments, but if `is-layout-constrained` is always being output, then we should probably switch off the layout switcher.

enejb replied:
good call.

andrewserong reviewed (commented):
Just left a few layout-related comments. Unfortunately these are drive-by comments without actually testing the PR! ~This is quite possibly a silly question, but how do I enable the file upload field block in a local environment? (I've tested it successfully in Dotcom simple sites before, but I couldn't quite wrap my head around testing locally and couldn't see anywhere in the dashboard for enabling beta blocks).~

~Sorry if it's an obvious flag to enable for testing locally, and I very well likely have made a mistake in my testing environment!~

Edit: I was using the Code Snippets plugin with `add_filter( 'jetpack_blocks_variation', function () { return 'beta'; } );` but I forgot to _enable_ the snippet 🤦

ramonjd committed:
Forms: fix inconsistent weights in outline style and terms labels (#43545)

* Ensure that existing font-weight styles do not override block and global label styles.

* Consent weights

* Revert options positioning kram

* Increase specificity to 0 1 0 for classic themes

* Bump specificity of consent label

* Bump specificity

andrewserong commented:
So here's a quick screenshot of what the default state of the block looks like with Flex and centered:

<img width="1157" alt="image" src="https://github.com/user-attachments/assets/29251e4c-a45b-4720-8887-21fa630f20c3" />

Perhaps something like that could work? We could then hide the layout switcher, but retain orientation? The main challenge from looking at this PR is that the layout block support doesn't "just" work out of the box the way it would for a "normal" block due to the shortcode rendering in Jetpack Forms blocks.

A couple of options:

* Restrict what we're displaying in the UI for controls even further, and only output the `is-layout-flex` and associated justification classnames where desired. OR:
* Call out to the `wp_render_layout_support_flag` function to generate the "real" layout styles and inject the appropriate classnames. This would be closely resemble what happens for a "normal" block that opts-in to layout, but might be more than what's needed for the requirements of this block.

Another question:

While testing this out, I found I couldn't click on the paragraph within the dropzone within the editor to change the text, as the dropzone itself becomes selected. Is that a known issue?

andrewserong commented:
Oh, one other note for if you're looking at using the core `wp_render_layout_support_flag` function. In the inner blocks refactor, the `get_block_support_classes_and_styles` function is currently responsible for calling out to the other block supports, so that could be a place to try out calling that function. Here's where the existing calls happen:

https://github.com/Automattic/jetpack/blob/85bb6edc9cd5c6060c5c55f3fa6b49e2bfd768b5/projects/packages/forms/src/contact-form/class-contact-form-plugin.php#L343-L347

talldan committed:
Forms: Fix animated style label font scaling (#43615)

* Fix animated style label font size in editor

* Add span wrapper around label text

* Update animated label styles on frontend

* Size and animation adjustments for required text

* Fix label animating when selected in editor if using the animated form style variation and no placeholder is set

talldan committed:
Forms Fix animated label centered positioning (#43635)

* Fix animated label vertical centering when using uneven borders in the input

* Remove unused cass var

enejb committed:
Add locking and prevent any special blocks from being added

enejb committed:
Use the constrained layout

enejb committed:
Move to the correct folder

talldan committed:
Forms: Remove options text color support (#43612)

* Remove text color support for jetpack/options, since frontend does not replicate the markup of the editor and so style parity cannot be achieved

* Try to fix tests

* Whoops, wrong flag

* One more time

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Forms: options block layout tweaks (#43531)

* Refactor contact form styles and rendering logic to enhance alignment and maintainability. Update CSS for radio and checkbox options to use consistent alignment, and streamline fieldset and div class attributes by removing unnecessary prefixes. Improve handling of border styles and attributes for outlined forms, ensuring compatibility with global styles.

* Try baseline for the options

* Revert changes to "fix" option alignment

* Revert option alignment v2

* Add classes to tests

* Reinstate tweaks for option positioning

* Apply wp-block-jetpack-option class in the attribute migration method

* Use em instead of rem. Remove unnecessary tweaks

* Border support for classic themes

* Ensure radio inputs don't warp in classic themes. Update options description

* Force opinionated dimensions on the options inputs

* wp-block-jetpack-option was not being applied to inner options blocks

ramonjd committed:
Forms: update padding for jetpack options in contact form to conform with trunk (#43634)

* consolidate legends

* Remove styles that are inconsistent with existing trunk spacing. (frontend)

* Remove padding in editor

* editor

* The feature branch has different markup for .wp-block-jetpack-options so we need to target the fieldset using different selectors

* comments

talldan committed:
Fix date field label suffix font size for outlined/animated styles

talldan committed:
Fix option input color not inherited in editor

ramonjd committed:
Forms: Update border radius and width handling in deprecateFieldStyles function (#43653)

* Attempt to fix warnings about PhanUndeclaredTypeParameter for WP_Block

* Forms: Update border radius and width handling in deprecateFieldStyles function

* Ensure border radius and width are correctly set to 'undefined' only when they are not provided or are null, allowing for proper handling of zero values.

* Coerce lineheight to str

* REvert  - bad rebase

* Update class-contact-form-plugin.php

My IDE auto formatted this

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Remove duplicated styles

enejb committed:
limit the blocks

enejb committed:
Make the animated and outline styles work

simison commented on projects/packages/forms/src/blocks/field-file/edit.js:
```diff
@@ -0,0 +1,64 @@
+	useFormWrapper( { attributes, clientId, name } );
+	const { blockStyle } = useJetpackFieldStyles( attributes );
+
+	let classes = clsx( className, 'jetpack-field is-non-animated-label', {
+		'is-selected': isSelected,
+	} );
+	if ( width ) {
+		classes += ` jetpack-field__width-${ width }`;
+	}
```
Untested but suppose this should work?

```suggestion
	let classes = clsx( className, 'jetpack-field is-non-animated-label', {
		'is-selected': isSelected,
		`jetpack-field__width-${ width }`: width,
	} );
```

enejb replied:
I tried this but it didn't work.

simison replied:
Then no biggie, current way is fine.

simison replied:
Or it might need brackets to work 🤔  
```js
[ `jetpack-field__width-${ width }` ]: width,
```

enejb committed:
Remove templateLock

ramonjd committed:
Form: fix default fallbacks for unmigrated forms (#43676)

* Outlined multi option backgrounds defaults

* Don't concat the label and option styles for consent and single option in unmigrated forms

* PHPCS warning - Use statements are not alphabetically sorted (MediaWiki.Classes.UnsortedUseStatements.UnsortedUse

ramonjd committed:
Forms: tame select arrow positioning (#43674)

* initial commit - adding extra markup to the select rendering and fixing positioning of labels

* uncommenting

* Remove default select border

* Update select field rendering to include a wrapper div with the correct class structure for improved styling and consistency in tests.

* Update contact form styles to inherit text color and adjust z-index for select wrapper elements

* Only run `getAttribute` on select wrapper if there is one

* Reinstate padding change

* phan me

* phanos

* whitespace

andrewserong commented on projects/packages/forms/src/blocks/dropzone/index.js:
```diff
@@ -0,0 +1,108 @@
+					justifyContent: 'center',
+					orientation: 'vertical',
+				},
+				spacing: {
+					padding: {
+						top: '48px',
+						bottom: '48px',
+						left: '48px',
+						right: '48px',
+					},
+					margin: {
+						top: '8px',
+						bottom: '8px',
+					},
+				},
+				border: {
+					style: 'dashed',
+					width: '1px',
+					color: 'rgba(125,125,125,0.3)',
+				},
+			},
```
Unfortunately, injecting these style values injects them at the individual block level which prevents global styles from working. Here's a screenshot of attempting to set an orange border within global styles, but the block level attributes are taking precedence:

<img width="1181" alt="image" src="https://github.com/user-attachments/assets/c03a0697-72d9-46b4-800a-be4d9fdf4c40" />

Could an alternative be to provide low-specificity CSS defaults of the styling that you want to as a baseline for this block, and then themes (and individual block markup) can override via the block supports if and when they need to?

This could be a neat alternative to the issue of picking the layout type, too, if there are still any outstanding issues with the layout type.

andrewserong reviewed (commented):
Just took this for another quick spin — I think there's a problem with injecting the default block attribute styles for the block as it essentially means the styles are baked in at the individual block level within post content, which takes precedence over global styles.

I'm wondering if it's better to provide base, low specificity defaults (i.e. 0,1,0 or similar) so the there's an opinionated base style for the block, but that themes and individual blocks can override it as needed.

Another alternative is [to filter the `theme.json` in PHP](https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/) to provide the styles you want as defaults, but depending on what you're doing, that might be more complex than simply providing base CSS.

Edit: also, is there an issue with the block rendering? After checking out the PR again, I'm seeing it in the editor, but not being rendered on the site frontend 🤔

enejb committed:
Default center blocks inside the dropzone but allow content justification control

enejb committed:
Fixes

enejb committed:
minro fix

enejb committed:
Improved focus management

enejb committed:
Add keyboard events for the remove button

ramonjd committed:
Ensure that animated labels sit above active input elements.

simison committed:
Remove link from the dropzone

enejb committed:
Update the dropzone icon.

enejb committed:
Add nicer file upload preview with icons

ramonjd committed:
Forms: account for missing placeholder when transforming input fields to select dropdown and fix z-index (#43721)

* Enhance select field transformation by adding default placeholder for input block. Update contact form plugin to correctly reference placeholder attribute. Adjust z-index values in CSS for improved label positioning.

* Update z-index values in contact form CSS for improved label positioning and remove redundant styles for outlined fields.

* Complete CSS comment

* Refactor select field transformation to set placeholder attribute directly from attributes. Remove redundant placeholder handling in inner block creation.

* Revert createTextFieldInnerBlocks

* Revert experiment. Add default placeholder for select in createTextFieldInnerBlocks

* Don't mutate the attributes object

* Don't spread optional chaining just in case

ramonjd committed:
Update contact form styles to prevent color overrides by global styles. Add 'color: unset !important;' to both editor.scss and grunion.css for consistent input appearance. (#43737)

enejb committed:
Update the icon

enejb committed:
Revert test files. Since we didn't add any new tests.

enejb committed:
Make it work nice for dark themes and icon

enejb committed:
Update the to nicer styles across themes
